File: src/main/java/org/jcodec/containers/mp4/boxes/TrackFragmentHeaderBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Track fragment header box
 * 
 * @author The JCodec project
 * 
 */
public class TrackFragmentHeaderBox extends FullBox {
    public TrackFragmentHeaderBox(Header atom) {
        super(atom);
    }

    //@formatter:off
    public static final int FLAG_BASE_DATA_OFFSET = 0x01;
    public static final int FLAG_SAMPLE_DESCRIPTION_INDEX = 0x02;
    public static final int FLAG_DEFAILT_SAMPLE_DURATION = 0x08;
    public static final int FLAG_DEFAULT_SAMPLE_SIZE = 0x10;
    public static final int FLAG_DEFAILT_SAMPLE_FLAGS = 0x20;
    //@formatter:on

    private int trackId;
    private long baseDataOffset;
    private int sampleDescriptionIndex;
    private int defaultSampleDuration;
    private int defaultSampleSize;
    private int defaultSampleFlags;

    public static String fourcc() {
        return "tfhd";
    }

    public static TrackFragmentHeaderBox createTrackFragmentHeaderBox(int trackId, long baseDataOffset,
            int sampleDescriptionIndex, int defaultSampleDuration, int defaultSampleSize, int defaultSampleFlags) {
        TrackFragmentHeaderBox box = new TrackFragmentHeaderBox(new Header(fourcc()));
        box.trackId = trackId;
        box.baseDataOffset = baseDataOffset;
        box.sampleDescriptionIndex = sampleDescriptionIndex;
        box.defaultSampleDuration = defaultSampleDuration;
        box.defaultSampleSize = defaultSampleSize;
        box.defaultSampleFlags = defaultSampleFlags;
        return box;
    }

    public static Factory create(int trackId) {
        return new Factory(TrackFragmentHeaderBox.createTrackFragmentHeaderBoxWithId(trackId));
    }

    public static Factory copy(TrackFragmentHeaderBox other) {
        TrackFragmentHeaderBox box = TrackFragmentHeaderBox
                .createTrackFragmentHeaderBox(other.trackId, other.baseDataOffset, other.sampleDescriptionIndex, other.defaultSampleDuration, other.defaultSampleSize, other.defaultSampleFlags);
        box.setFlags(other.getFlags());
        box.setVersion(other.getVersion());
        return new Factory(box);
    }

    public static TrackFragmentHeaderBox createTrackFragmentHeaderBoxWithId(int trackId) {
        TrackFragmentHeaderBox box = new TrackFragmentHeaderBox(new Header(fourcc()));
        box.trackId = trackId;
        return box;
    }

    public static class Factory {

        private TrackFragmentHeaderBox box;

        public Factory(TrackFragmentHeaderBox box) {
            this.box = box;
        }

        public Factory baseDataOffset(long baseDataOffset) {
            box.flags |= FLAG_BASE_DATA_OFFSET;
            box.baseDataOffset = (int) baseDataOffset;
            return this;
        }

        public Factory sampleDescriptionIndex(long sampleDescriptionIndex) {
            box.flags |= FLAG_SAMPLE_DESCRIPTION_INDEX;
            box.sampleDescriptionIndex = (int) sampleDescriptionIndex;
            return this;
        }

        public Factory defaultSampleDuration(long defaultSampleDuration) {
            box.flags |= FLAG_DEFAILT_SAMPLE_DURATION;
            box.defaultSampleDuration = (int) defaultSampleDuration;
            return this;
        }

        public Factory defaultSampleSize(long defaultSampleSize) {
            box.flags |= FLAG_DEFAULT_SAMPLE_SIZE;
            box.defaultSampleSize = (int) defaultSampleSize;
            return this;
        }

        public Factory defaultSampleFlags(long defaultSampleFlags) {
            box.flags |= FLAG_DEFAILT_SAMPLE_FLAGS;
            box.defaultSampleFlags = (int) defaultSampleFlags;
            return this;
        }

        public TrackFragmentHeaderBox create() {
            try {
                return box;
            } finally {
                box = null;
            }
        }
    }

    @Override
    public void parse(ByteBuffer input) {
        super.parse(input);
        trackId = input.getInt();
        if (isBaseDataOffsetAvailable())
            baseDataOffset = input.getLong();
        if (isSampleDescriptionIndexAvailable())
            sampleDescriptionIndex = input.getInt();
        if (isDefaultSampleDurationAvailable())
            defaultSampleDuration = input.getInt();
        if (isDefaultSampleSizeAvailable())
            defaultSampleSize = input.getInt();
        if (isDefaultSampleFlagsAvailable())
            defaultSampleFlags = input.getInt();
    }

    @Override
    protected void doWrite(ByteBuffer out) {
        super.doWrite(out);
        out.putInt(trackId);
        if (isBaseDataOffsetAvailable())
            out.putLong(baseDataOffset);
        if (isSampleDescriptionIndexAvailable())
            out.putInt(sampleDescriptionIndex);
        if (isDefaultSampleDurationAvailable())
            out.putInt(defaultSampleDuration);
        if (isDefaultSampleSizeAvailable())
            out.putInt(defaultSampleSize);
        if (isDefaultSampleFlagsAvailable())
            out.putInt(defaultSampleFlags);
    }
    
    @Override
    public int estimateSize() {
        return 40;
    }

    public int getTrackId() {
        return trackId;
    }

    public long getBaseDataOffset() {
        return baseDataOffset;
    }

    public int getSampleDescriptionIndex() {
        return sampleDescriptionIndex;
    }

    public int getDefaultSampleDuration() {
        return defaultSampleDuration;
    }

    public int getDefaultSampleSize() {
        return defaultSampleSize;
    }

    public int getDefaultSampleFlags() {
        return defaultSampleFlags;
    }

    public boolean isBaseDataOffsetAvailable() {
        return (flags & FLAG_BASE_DATA_OFFSET) != 0;
    }

    public boolean isSampleDescriptionIndexAvailable() {
        return (flags & FLAG_SAMPLE_DESCRIPTION_INDEX) != 0;
    }

    public boolean isDefaultSampleDurationAvailable() {
        return (flags & FLAG_DEFAILT_SAMPLE_DURATION) != 0;
    }

    public boolean isDefaultSampleSizeAvailable() {
        return (flags & FLAG_DEFAULT_SAMPLE_SIZE) != 0;
    }

    public boolean isDefaultSampleFlagsAvailable() {
        return (flags & FLAG_DEFAILT_SAMPLE_FLAGS) != 0;
    }

    public void setTrackId(int trackId) {
        this.trackId = trackId;
    }

    public void setDefaultSampleFlags(int defaultSampleFlags) {
        this.defaultSampleFlags = defaultSampleFlags;
    }

    public static TrackFragmentHeaderBox createTrackFragmentHeaderBox() {
        return new TrackFragmentHeaderBox(new Header(fourcc()));
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/TrackFragmentHeaderBox.kt
package org.jcodec.containers.mp4.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Track fragment header box
 *
 * @author The JCodec project
 */
class TrackFragmentHeaderBox(atom: Header) : FullBox(atom) {
    //@formatter:on
    private var trackId = 0
    var baseDataOffset: Long = 0
        private set
    var sampleDescriptionIndex = 0
        private set
    var defaultSampleDuration = 0
        private set
    var defaultSampleSize = 0
        private set
    private var defaultSampleFlags = 0

    class Factory(private var box: TrackFragmentHeaderBox?) {
        fun baseDataOffset(baseDataOffset: Long): Factory {
            box!!.flags = box!!.flags or FLAG_BASE_DATA_OFFSET
            box!!.baseDataOffset = baseDataOffset
            return this
        }

        fun sampleDescriptionIndex(sampleDescriptionIndex: Long): Factory {
            box!!.flags = box!!.flags or FLAG_SAMPLE_DESCRIPTION_INDEX
            box!!.sampleDescriptionIndex = sampleDescriptionIndex.toInt()
            return this
        }

        fun defaultSampleDuration(defaultSampleDuration: Long): Factory {
            box!!.flags = box!!.flags or FLAG_DEFAILT_SAMPLE_DURATION
            box!!.defaultSampleDuration = defaultSampleDuration.toInt()
            return this
        }

        fun defaultSampleSize(defaultSampleSize: Long): Factory {
            box!!.flags = box!!.flags or FLAG_DEFAULT_SAMPLE_SIZE
            box!!.defaultSampleSize = defaultSampleSize.toInt()
            return this
        }

        fun defaultSampleFlags(defaultSampleFlags: Long): Factory {
            box!!.flags = box!!.flags or FLAG_DEFAILT_SAMPLE_FLAGS
            box!!.defaultSampleFlags = defaultSampleFlags.toInt()
            return this
        }

        fun create(): TrackFragmentHeaderBox? {
            return try {
                box
            } finally {
                box = null
            }
        }

    }

    override fun parse(input: ByteBuffer) {
        super.parse(input)
        trackId = input.int
        if (isBaseDataOffsetAvailable) baseDataOffset = input.long
        if (isSampleDescriptionIndexAvailable) sampleDescriptionIndex = input.int
        if (isDefaultSampleDurationAvailable) defaultSampleDuration = input.int
        if (isDefaultSampleSizeAvailable) defaultSampleSize = input.int
        if (isDefaultSampleFlagsAvailable) defaultSampleFlags = input.int
    }

    override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putInt(trackId)
        if (isBaseDataOffsetAvailable) out.putLong(baseDataOffset)
        if (isSampleDescriptionIndexAvailable) out.putInt(sampleDescriptionIndex)
        if (isDefaultSampleDurationAvailable) out.putInt(defaultSampleDuration)
        if (isDefaultSampleSizeAvailable) out.putInt(defaultSampleSize)
        if (isDefaultSampleFlagsAvailable) out.putInt(defaultSampleFlags)
    }

    override fun estimateSize(): Int {
        return 40
    }

    fun getTrackId(): Int {
        return trackId
    }

    fun getDefaultSampleFlags(): Int {
        return defaultSampleFlags
    }

    val isBaseDataOffsetAvailable: Boolean
        get() = flags and FLAG_BASE_DATA_OFFSET != 0

    val isSampleDescriptionIndexAvailable: Boolean
        get() = flags and FLAG_SAMPLE_DESCRIPTION_INDEX != 0

    val isDefaultSampleDurationAvailable: Boolean
        get() = flags and FLAG_DEFAILT_SAMPLE_DURATION != 0

    val isDefaultSampleSizeAvailable: Boolean
        get() = flags and FLAG_DEFAULT_SAMPLE_SIZE != 0

    val isDefaultSampleFlagsAvailable: Boolean
        get() = flags and FLAG_DEFAILT_SAMPLE_FLAGS != 0

    fun setTrackId(trackId: Int) {
        this.trackId = trackId
    }

    fun setDefaultSampleFlags(defaultSampleFlags: Int) {
        this.defaultSampleFlags = defaultSampleFlags
    }

    companion object {
        //@formatter:off
        const val FLAG_BASE_DATA_OFFSET = 0x01
        const val FLAG_SAMPLE_DESCRIPTION_INDEX = 0x02
        const val FLAG_DEFAILT_SAMPLE_DURATION = 0x08
        const val FLAG_DEFAULT_SAMPLE_SIZE = 0x10
        const val FLAG_DEFAILT_SAMPLE_FLAGS = 0x20
        @JvmStatic
        fun fourcc(): String {
            return "tfhd"
        }

        fun tfhd(trackId: Int, baseDataOffset: Long, sampleDescriptionIndex: Int, defaultSampleDuration: Int, defaultSampleSize: Int, defaultSampleFlags: Int): TrackFragmentHeaderBox {
            val box = TrackFragmentHeaderBox(Header(fourcc()))
            box.trackId = trackId
            box.baseDataOffset = baseDataOffset
            box.sampleDescriptionIndex = sampleDescriptionIndex
            box.defaultSampleDuration = defaultSampleDuration
            box.defaultSampleSize = defaultSampleSize
            box.defaultSampleFlags = defaultSampleFlags
            return box
        }

        fun create(trackId: Int): Factory {
            return Factory(createTrackFragmentHeaderBoxWithId(trackId))
        }

        fun copy(other: TrackFragmentHeaderBox): Factory {
            val box = tfhd(other.trackId, other.baseDataOffset, other.sampleDescriptionIndex, other.defaultSampleDuration, other.defaultSampleSize, other.defaultSampleFlags)
            box.flags = other.flags
            box.version = other.version
            return Factory(box)
        }

        fun createTrackFragmentHeaderBoxWithId(trackId: Int): TrackFragmentHeaderBox {
            val box = TrackFragmentHeaderBox(Header(fourcc()))
            box.trackId = trackId
            return box
        }

        fun createTrackFragmentHeaderBox(): TrackFragmentHeaderBox {
            return TrackFragmentHeaderBox(Header(fourcc()))
        }
    }
}